diff --git a/source/Backend/UserMode/ExecMemory/code-patch-tool-posix.cc b/source/Backend/UserMode/ExecMemory/code-patch-tool-posix.cc
index bfe7361..ea32122 100644
--- a/source/Backend/UserMode/ExecMemory/code-patch-tool-posix.cc
+++ b/source/Backend/UserMode/ExecMemory/code-patch-tool-posix.cc
@@ -11,7 +11,7 @@ PUBLIC int DobbyCodePatch(void *address, uint8_t *buffer, uint32_t buffer_size)
 #if defined(__ANDROID__) || defined(__linux__)
   int page_size = (int)sysconf(_SC_PAGESIZE);
   uintptr_t patch_page = ALIGN_FLOOR(address, page_size);
-  uintptr_t patch_end_page = ALIGN_FLOOR((uintptr_t)address + buffer_size, page_size);
+  uintptr_t patch_end_page = ALIGN_FLOOR((uintptr_t)address + buffer_size - 1, page_size);
 
   // change page permission as rwx
   mprotect((void *)patch_page, page_size, PROT_READ | PROT_WRITE | PROT_EXEC);
